010-anonymous-login.md

```diff
@@ -0,0 +1,114 @@
+- Start Date: 15-02-2021
+- Target Date: 16-02-2021
+- Appwrite Issue: https://github.com/appwrite/appwrite/discussions/907
+
+## Summary
+
+[summary]: #summary
+
+This RFC will introduce authentication of anonymous users in Appwrite. These temporary anonymous users can be used to allow users who haven't yet signed up to an app to work with data protected by Appwrite's permission and security rules. If an anonymous user decides to sign up to an app, they can convert their account to an E-Mail or OAuth2 authenticated one, so that they can continue to work with their protected data in future sessions.
+
+## #Problem Statement (Step 1)
+
+[problem-statement]: #problem-statement
+
+**What problem are you trying to solve?**
+
+When you develop an application, there might be times when you want to let a user interact with pieces of Appwrite before they're signed in.
+
+**What is the context or background in which this problem exists?**
+
+For example, say you're an online retailer, you might want to let users browse your inventory and place items in a shopping cart. As the retailer, you don't know who the user is, so they're anonymous. A user might remain anonymous if they never choose to sign in. If they do choose to sign in, then the anonymous becomes a known user and their anonynmous account is converted from anonymous to a user account.
+
+**Once the proposal is implemented, how will the system change?**
+
+Appwrite will offer a project wide feature to enable anonymous logins.
+
+## Design proposal (Step 2)
+
+[design-proposal]: #design-proposal
+
+An anonymous user will be treated like a unverified normal user with following limitations:
+
+- No Name
+- No E-Mail address
+
+The user can convert their account to one, that is authenticated via E-Mail or an OAuth2 provider. To convert an account to one that is authenticated with an E-Mail address, we should offer an endpoint similiar to the `/v1/account/verification` one. Meaning, an accouunt can only be converted by clicking a verification link form an E-Mail. This way we ensure an anonymous account doesn't get lost with a typo in the E-Mail address. The flow should look like this:
+
+- Anonymous user initiate the E-Mail verification process
+- Verifies his E-Mail via the link provided by the E-mail
+- Will only be converted to a normal user at this point.
+
+The process for OAuth2 should be very simple, once a user has linked an OAuth2 authentification to an anonymous account - it will be converted.
+
+For data consistency, the user should be provided the same ID at any point.
+
+### Endpoints
+
+#### `GET: /v1/account/sessions/anonymous`
+
+Allow a user to create and login to an anonymous user account.
+
+#### `POST: /v1/account/verification/anonymous`
+
+Allow a user to initiate converting their account to one that is authenticated via E-Mail. This will have the same functionality as the `POST/v1/account/verification` endpoint, except that an E-Mail address needs to be provided in the payload.
```

Comment:
What happens if in the future will offer account authentication using other methods like SMS? Should we consider allowing anonymous users to use the `PATCH /v1/account/email` using an empty password and then optionally allow app makers to start verification using the existing verification flow?

With this suggestion we can better reuse endpoint and make future verification methods easier to implement

Reply:
Yes, that should work 👍🏻

Reply:
@eldadfux having some doubts about this now. Anonymous users would need to do 2 API calls (setting email and password) before even being able to initiate this verification flow. We would probably need to have an endpoint to pass the E-Mail and Password to start the verification.

Reply:
Actually this has proved itself very good for us when we separated account creation and login for REST compatibility. As I see it this is a limitation of the REST protocol, but in this case I would favor flexibility over performance, each developer should decide its own app flow, and the 50ms penalty shouldn't be too much fo this kind of process.

Also, creating multiple endpoint just for small performance gains will open the door for us to introduce a lot of added complexity on the Appwrite API.

Reply:
As discussed, this will be part of the `PATCH: /v1/account/email` endpoint now.